packages/common/src/hardhat/tasks/xchainSetup.ts

```diff
@@ -253,21 +275,239 @@ task("setup-l2-xchain", "Configures L2 cross chain smart contracts").setAction(a
 
   const Finder = await deployments.get("Finder");
   const finder = new web3.eth.Contract(Finder.abi, Finder.address);
-  const chainId = await getChainId();
-  let ChildMessenger;
-  switch (chainId) {
-    case "42161":
-      ChildMessenger = await deployments.get("Arbitrum_ChildMessenger");
-      break;
-    case "288":
-      ChildMessenger = await deployments.get("Boba_ChildMessenger");
-      break;
-    default:
-      throw new Error("Unimplemented L2");
-  }
+  const chainId = Number(await getChainId());
+  const chainName = getChainNameForId(chainId);
+  const ChildMessenger = await deployments.get(`${chainName}_ChildMessenger`);
   const Registry = await deployments.get("Registry");
 
   console.log(`Found Finder @ ${finder.options.address}`);
 
   await setupChildMessenger(finder, deployer, ChildMessenger, Registry);
 });
+
+task("verify-xchain", "Checks ownership state of cross chain smart contracts")
+  .addParam("l2", "Chain name of the child messenger to check")
+  .setAction(async function (taskArguments, hre_) {
+    const hre = hre_ as CombinedHRE;
+    const { deployments, web3, companionNetworks } = hre;
+    const { l2 } = taskArguments;
+
+    assert(Object.keys(L2_CHAIN_NAMES).includes(l2), "Invalid L2 chain name");
+    // Each parent/child messenger has a different setup, so alert user if their specified L2 is implemented in this verification script. For example,
+    // Polygon might be included L2_CHAIN_NAMES but its specific setup is not yet implemented in this verification script.
+    if (!["arbitrum", "optimism", "boba"].includes(l2)) throw new Error("Unimplemented verification for l2 chain name");
+    const l2ChainId = getChainIdForName(l2);
+    const l2ChainName = getChainNameForId(l2ChainId);
+    assert(process.env[`NODE_URL_${l2ChainId}`], `Must set NODE_URL_${l2ChainId} in the environment`);
+    const l1Web3 = web3;
+    const l2Web3 = new Web3(process.env[`NODE_URL_${l2ChainId}`] as string);
+
+    const [
+      governorHub,
+      oracleHub,
+      governor,
+      parentMessenger,
+      childMessenger,
+      oracleSpoke,
+      governorSpoke,
+      l1Registry,
+      l2Registry,
+      l2Store,
+      l2IdentifierWhitelist,
+      l2AddressWhitelist,
+      l2OptimisticOracle,
+      l2Finder,
+    ] = await Promise.all([
+      deployments.get("GovernorHub"),
+      deployments.get("OracleHub"),
+      deployments.get("Governor"),
+      deployments.get(`${l2ChainName}_ParentMessenger`),
+      companionNetworks[l2ChainName.toLowerCase()].deployments.get(`${l2ChainName}_ChildMessenger`),
```

Comment:
might it be useful to define a `companionNetworkGet = companionNetworks[l2ChainName.toLowerCase()].deployments.get` to make this a bit cleaner?

Reply:
will do